src/v/cluster/cloud_data/aggregated_uploader.h

```diff
@@ -0,0 +1,436 @@
+                // Signal EOS
+                co_return ss::temporary_buffer<char>();
+            }
+            // This iobuf stores placeholders that reference the data
+            data_layout layout;
+            // Convert reader to iobuf
+            auto& [ntp, rdr] = _readers.at(_refill_ix++);
+            co_await rdr.consume(
+              serializing_consumer(ntp, &_buff, _layout),
+              model::timeout_clock::now() + std::chrono::seconds(10));
+            vassert(!_buff.empty(), "Reader returned empty buffer");
+        }
+        // Invariant: _buff has some data
+        auto b = _buff.begin()->share();
+        _buff.pop_front();
+        co_return std::move(b);
+    }
+
+private:
+    // Move readers from the input map (ord. by ntp) into the _readers
+    // collection.
+    // Data is not copied. Only the record_batch_reader instances are
+    // reshuffled.
+    void linearize(ss::foreign_ptr<aggregated_write_request_ptr>& inp) {
+        for (auto& [ntp, requests] : inp->to_upload) {
+            // TODO: add delimiter here
+            // create a model::record_batch_reader for an individual record
+            // batch that contains the delimiter (introduce the new batch type)
+            for (auto& req : requests) {
+                _readers.push_back(std::make_tuple(
+                  std::make_optional(ntp), std::move(req->reader)));
+            }
+        }
+    }
+
+    // We're storing ntp + reader for normal batches and nullopt + reader for
+    // delimiters and metadata batches.
+    chunked_vector<
+      std::tuple<std::optional<model::ntp>, model::record_batch_reader>>
+      _readers;
+    size_t _refill_ix{0};
+    iobuf _buff;
+    data_layout* _layout;
+};
+
+} // namespace details
+
+/// The data path uploader
+///
+/// This is a peering service. All ingested data on every
+/// shard is going though this service.
+///
+/// Every shard collects a list of write_request instances in
+/// memory. Periodically, the data is uploaded to the cloud storage
+/// and removed from memory.
+///
+/// Every shard is a queue. The requests are added to the end of the queue
+/// and the 0-shard consumes them from the beginning. The queue is passive.
+/// All operations are started by shard 0 which simplifies interactions
+/// between the components. Shard 0 performs the following operations:
+/// - pushes expiration time, shards react by failing all write_requests with
+///   expiration_time below this value.
+/// - consumes write_requests from the _pending list. The request is
+///   'consumed' when shard 0 created a placeholder for it (using c-tor(2) of
+///   the write_request)
+/// - starts upload on one of the shards (one at a time). The
+///   aggregated_uploader on the shard is still passive. The shard 0 is doing
+///   the work.
+///
+/// Shard 0 works as both controller and a queue.
+/// Every shard has an epoch counter. Every new request is getting a unique
```

Comment:
Please explain the motivation behind this epoch counter. Btw, it reads to me more like `upload request id` rather than epoch. Epoch as a name is used when it describes a collection of things rather than one discrete element.